fe/src/main/coffee/bustracker.coffee

```diff
@@ -0,0 +1,226 @@
+            map: map
+  
+        newLatenessSeconds = if (age = toPosition.age) and oldLatenessSeconds
+          Math.max oldLatenessSeconds, age
+        else
+          oldLatenessSeconds or age
+  
+        [
+          newLatenessSeconds
+          oldIsDubious or not normalBounds.contains toLl
+          [oldLines..., line]
+          [oldMarkers..., marker]
+          map
+        ]
+  
+      @render = (routeUnsorted, map) ->
+        route = routeUnsorted.sort((r0, r1) -> r1.timestamp.localeCompare r0.timestamp)
+        routeFromTo = zip [undefined, route...], route
+        relevantRouteFromTo = routeFromTo.slice 0, maxTrail
+        initState = [0, false, [], [], map]
+        finalState = relevantRouteFromTo.reduce drawLineBetweenMarkers, initState
+        [latenessSeconds, isDubious, newLines, newMarkers, _] = finalState
+        warnAboutDelayedBuses latenessSeconds
+        removeOldLinesAndMarkers()
+  
+        # TODO can we get rid of the destructive assignation here?
+        lines = newLines
+        markers = newMarkers
+  
+        [isDubious, relevantRouteFromTo.length]
+  
+  black = new Route 'black', '#111'
+  red = new Route 'red', '#D11'
+  green = new Route 'green', '#1D1'
+  
+  (phoneId) ->
+    switch phoneId
+      when "HTC Desire C" then black
+      when "GT-I8190N" then black
+      when "blackberry" then black
+      when "HTC Desire S" then red
+      when "strawberry" then red
+      else green
+
+getRouteForPhone = getGetRouteForPhone()
+
+normalZone = new google.maps.Rectangle
+  strokeColor: '#22DD22'
+  strokeOpacity: 0.7
+  strokeWeight: 2
+  fillColor: '#22DD22'
+  fillOpacity: 0.1
+  bounds: normalBounds
+
+stationX = 51.76373
+stationY = -0.215564
+titanX = 51.762488
+titanY = -0.243518
+centerX = 51.764
+centerY = -0.230
+
+#configuration
+maxTrail = 5
+jsonRefreshInterval = 11500
+defaultZoom = 14
+maxLat = 51.7757
+minLat = 51.755117
+maxLng = -0.208826
+minLng = -0.251999
+maxLatenessSeconds = 60
+#minLng = -0.235 # for testing
+
+#magic numbers
+msPerSecond = 1000
+secondsPerMinute = 60
+secondsPerHour = 3600
+minutesPerHour = 60
+
+llStation = new LatLng stationX, stationY
+llTitan = new LatLng titanX, titanY
+center = new LatLng centerX, centerY
+bl = new LatLng minLat, minLng
+tr = new LatLng maxLat, maxLng
+normalBounds = new LatLngBounds bl, tr
+
+handleMissingAndDubiousBusInfo = (isDubious, isMissing, map) ->
+  if isMissing
+    $("#missing-bus-info").addClass("show").removeClass("hide")
+  else
+    $("#missing-bus-info").addClass("hide").removeClass("show")
+    if isDubious
+      normalZone.setMap map
+      $("#one-is-outside").addClass("show").removeClass("hide")
+    else
+      normalZone.setMap null
+      $("#one-is-outside").addClass("hide").removeClass("show")
+
+#polyfill
+zip = (left, right) -> [left[i], a] for a, i in right
+
+jsonHdlr = (routes, map, oldValidity, dataUrl) ->
+
+  $('#last-checked').text(new Date)
+
+  warnings = for phoneId, routeContent of routes
```

Comment:
if `of` is considered bad practice, this can be replaced with an array of arrays and 
`for [phoneId, routeContent] of routes` can be done. We don't ever look up by key.

But I think this way is semantically clearer and makes the JSON easier to read